File: pipeline/columns.py
"""Typed-array column envelope: the Python half of app/src/schema/columns.ts.

Every column the pipeline writes goes through encode_column, which applies the dtype rule
(docs/decisions.md) and always writes little-endian bytes via explicit '<i4' / '<f4' dtypes,
so output never depends on the build machine's native byte order.

    counts, ids (incl. categorical codes)  → int32
    shares, rates, indices                 → float32
    money                                  → float32, unit "cad_millions"
    measure                                → float32, unit required
"""

from __future__ import annotations

import base64
import re
from typing import Any

import numpy as np

BYTE_ORDER = "le"
MONEY_UNIT = "cad_millions"
INT32_LE = np.dtype("<i4")
FLOAT32_LE = np.dtype("<f4")

KIND_DTYPE: dict[str, np.dtype] = {
    "count": INT32_LE,
    "id": INT32_LE,
    "share": FLOAT32_LE,
    "rate": FLOAT32_LE,
    "index": FLOAT32_LE,
    "money": FLOAT32_LE,
    "measure": FLOAT32_LE,
}
COLUMN_NAME_RE = re.compile(r"^[a-z][a-z0-9]*(?:_[a-z0-9]+)*$")
UNIT_RE = re.compile(r"^[a-z][a-z0-9_]*$")


def file_meta(**provenance: str | int | float | bool) -> dict[str, Any]:
    """File-level meta for any artefact carrying encoded columns."""
    return {"byteOrder": BYTE_ORDER, **provenance}


def check_column_name(name: str) -> str:
    if not COLUMN_NAME_RE.match(name):
        raise ValueError(f"column name {name!r} is not snake_case")
    return name


def encode_column(
    values: Any,
    kind: str,
    *,
    unit: str | None = None,
    method: str | None = None,
    confidence: float | None = None,
) -> dict[str, Any]:
    if kind not in KIND_DTYPE:
        raise ValueError(f"unknown column kind {kind!r}")
    target = KIND_DTYPE[kind]
    arr = np.asarray(values)
    if arr.ndim != 1:
        raise ValueError(f"columns are 1-D, got shape {arr.shape}")

    if target == INT32_LE:
        if arr.size and not np.issubdtype(arr.dtype, np.integer):
            raise TypeError(f"kind {kind!r} needs integer values, got {arr.dtype}; round explicitly first")
        info = np.iinfo(np.int32)
        if arr.size and (arr.min() < info.min or arr.max() > info.max):
            raise OverflowError(f"kind {kind!r} values exceed int32")
    elif arr.size and not (np.issubdtype(arr.dtype, np.floating) or np.issubdtype(arr.dtype, np.integer)):
        raise TypeError(f"kind {kind!r} needs numeric values, got {arr.dtype}")

    if kind == "money" and unit != MONEY_UNIT:
        raise ValueError(f"money columns must use unit {MONEY_UNIT!r}")
    if kind == "measure" and (unit is None or not UNIT_RE.match(unit)):
        raise ValueError("measure columns need a snake_case unit, e.g. 'km'")
    if kind not in ("money", "measure") and unit is not None:
        raise ValueError(f"kind {kind!r} takes no unit")
    if confidence is not None and not 0 <= confidence <= 1:
        raise ValueError("confidence must be in [0, 1]")

    column: dict[str, Any] = {
        "kind": kind,
        "dtype": "int32" if target == INT32_LE else "float32",
        "byteOrder": BYTE_ORDER,
        "length": int(arr.size),
        "data": base64.b64encode(arr.astype(target).tobytes()).decode("ascii"),
    }
    if unit is not None:
        column["unit"] = unit
    if method is not None:
        column["method"] = method
    if confidence is not None:
        column["confidence"] = confidence
    return column


def decode_column(column: dict[str, Any]) -> np.ndarray:
    if column.get("byteOrder") != BYTE_ORDER:
        raise ValueError(f"unsupported byteOrder {column.get('byteOrder')!r}")
    dtype = {"int32": INT32_LE, "float32": FLOAT32_LE}[column["dtype"]]
    raw = base64.b64decode(column["data"], validate=True)
    if len(raw) != column["length"] * 4:
        raise ValueError(f"byte length {len(raw)} does not match {column['length']} × 4")
    return np.frombuffer(raw, dtype=dtype)

File: app/src/schema/columns.ts
import { z } from 'zod';

/**
 * Wire encoding for typed-array columns shared by every Meridian artefact.
 *
 * JSON cannot carry a Float32Array, so columns travel as base64 of the raw bytes plus a
 * dtype and an explicit byte order. Bytes are always little-endian: the Python writer uses
 * '<i4' / '<f4' (pipeline/columns.py) and this decoder reads through DataView with
 * littleEndian = true, so neither side depends on the host's native order.
 *
 * The dtype rule (docs/decisions.md) is enforced here through `kind`:
 *   count, id               → int32
 *   share, rate, index      → float32
 *   money                   → float32, unit "cad_millions"
 *   measure                 → float32, unit required (e.g. "km")
 */

export const BYTE_ORDER = 'le' as const;
export const MONEY_UNIT = 'cad_millions' as const;

export const INT_KINDS = ['count', 'id'] as const;
export const FLOAT_KINDS = ['share', 'rate', 'index'] as const;
export type ColumnKind = (typeof INT_KINDS)[number] | (typeof FLOAT_KINDS)[number] | 'money' | 'measure';

/** snake_case: lowercase words of letters/digits joined by single underscores. */
export const ColumnNameSchema = z
  .string()
  .regex(/^[a-z][a-z0-9]*(?:_[a-z0-9]+)*$/)
  .meta({ id: 'ColumnName' });

const envelope = {
  byteOrder: z.literal(BYTE_ORDER).describe('Byte order of data; always little-endian'),
  length: z.int().nonnegative().describe('Element count, not byte count'),
  data: z.base64().describe('Base64 of the element bytes in byteOrder'),
  method: z
    .string()
    .optional()
    .describe('How the values were produced, e.g. "allocation_v1" for estimated GDP'),
  confidence: z
    .number()
    .min(0)
    .max(1)
    .optional()
    .describe('0–1; present on any estimated column so the UI can label it'),
};

export const EncodedColumnSchema = z
  .discriminatedUnion('kind', [
    z.strictObject({
      kind: z.enum(INT_KINDS).describe('Counts and ids (including categorical codes) are int32'),
      dtype: z.literal('int32'),
      ...envelope,
    }),
    z.strictObject({
      kind: z.enum(FLOAT_KINDS).describe('Shares, rates and indices are float32'),
      dtype: z.literal('float32'),
      ...envelope,
    }),
    z.strictObject({
      kind: z.literal('money').describe('Money is float32 in millions of CAD'),
      dtype: z.literal('float32'),
      unit: z.literal(MONEY_UNIT),
      ...envelope,
    }),
    z.strictObject({
      kind: z.literal('measure').describe('Physical measures are float32 with an explicit unit'),
      dtype: z.literal('float32'),
      unit: z.string().regex(/^[a-z][a-z0-9_]*$/),
      ...envelope,
    }),
  ])
  .meta({ id: 'EncodedColumn' });

/** An int32 id column on its own, for places that only accept ids (assignments, side tables). */
export const IdColumnSchema = z
  .strictObject({ kind: z.literal('id'), dtype: z.literal('int32'), ...envelope })
  .meta({ id: 'IdColumn' });

export type EncodedColumn = z.infer<typeof EncodedColumnSchema>;
export type TypedColumn = Float32Array | Int32Array;

/** File-level metadata carried by every artefact that contains encoded columns. */
export const FileMetaSchema = z
  .object({ byteOrder: z.literal(BYTE_ORDER) })
  .catchall(z.union([z.string(), z.number(), z.boolean()]))
  .describe('byteOrder plus free-form provenance, e.g. gdp_method')
  .meta({ id: 'FileMeta' });

export interface ColumnSpec {
  kind: ColumnKind;
  unit?: string;
  method?: string;
  confidence?: number;
}

export function dtypeForKind(kind: ColumnKind): EncodedColumn['dtype'] {
  return (INT_KINDS as readonly string[]).includes(kind) ? 'int32' : 'float32';
}

export function encodeColumn(values: TypedColumn, spec: ColumnSpec): EncodedColumn {
  const dtype = dtypeForKind(spec.kind);
  if ((dtype === 'int32') !== values instanceof Int32Array) {
    throw new TypeError(`kind "${spec.kind}" must be ${dtype}, got ${values.constructor.name}`);
  }
  const bytes = new Uint8Array(values.length * 4);
  const view = new DataView(bytes.buffer);
  for (let i = 0; i < values.length; i++) {
    if (dtype === 'int32') view.setInt32(i * 4, values[i], true);
    else view.setFloat32(i * 4, values[i], true);
  }
  let binary = '';
  for (let i = 0; i < bytes.length; i++) binary += String.fromCharCode(bytes[i]);
  return EncodedColumnSchema.parse({
    ...spec,
    dtype,
    byteOrder: BYTE_ORDER,
    length: values.length,
    data: btoa(binary),
  });
}

export function decodeColumn(column: EncodedColumn): TypedColumn {
  if (column.byteOrder !== BYTE_ORDER) throw new Error(`unsupported byteOrder ${column.byteOrder}`);
  const binary = atob(column.data);
  if (binary.length !== column.length * 4) {
    throw new Error(`Column byte length ${binary.length} does not match ${column.length} × 4`);
  }
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  const view = new DataView(bytes.buffer);
  if (column.dtype === 'float32') {
    const out = new Float32Array(column.length);
    for (let i = 0; i < out.length; i++) out[i] = view.getFloat32(i * 4, true);
    return out;
  }
  const out = new Int32Array(column.length);
  for (let i = 0; i < out.length; i++) out[i] = view.getInt32(i * 4, true);
  return out;
}

/** ISO 8601 calendar date. Years are four digits, so 1000-01-01 is valid. */
export const IsoDateSchema = z.iso.date().meta({ id: 'IsoDate' });
